### ISSUE TITLE
intruder_alarm stability

### ISSUE THREAD

**Issue by hawkw** [hawkw/alarm#12]
Right now, `core::ptr::NonNull` (née `ptr::Shared`) is the main unstable feature blocking `intruder_alarm` from building with the stable compiler. As of rust-lang/rust#46952, it's coming to stable Real Soon Now (under its' new name).

The other gated feature that `intruder_alarm` uses is [`const_fn`](https://github.com/hawkw/alarm/blob/master/intruder_alarm/src/lib.rs#L30), which is pretty easy to hack around (e.g. with a macro) or just with conditional compilation. In light of that, `intruder_alarm` may soon compile on stable Rust.

Let's try to avoid adding any new feature gates, so that when `ptr::NonNull` makes it to stable, `intruder_alarm` will build against stable.

**Comment by leshow** [comment]
Isn't the custom allocator stuff for nightly only also?

**Comment by hawkw** [comment]
@leshow 
> Isn't the custom allocator stuff for nightly only also?

Yeah, that's correct. But if nothing is keeping the intrusive collections crate --- separate from the actual allocator implementations --- from building on stable, why not? I'm sure there are *some* other use-cases for intrusive collections other than writing allocators...

**Comment by hawkw** [comment]
(which is my way of saying, "i'd like to ship `intruder_alarm` to crates.io independently of the allocators)
